File: Smodr/src/main/java/com/kevintcoughlin/smodr/services/MediaService.java
package com.kevintcoughlin.smodr.services;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.Binder;
import android.os.IBinder;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.kevintcoughlin.smodr.models.Item;

import org.jetbrains.annotations.Contract;

interface IMediaService {
    void seekTo(final int milliseconds);

    boolean isPlaying();

    int getDurationInMilliseconds();

    int currentPositionInMilliseconds();

    void resumePlayback();

    void pausePlayback();

    void stopPlayback();

    void forward();

    void rewind();

    void setPlaybackListener(final MediaService.IPlaybackListener listener);
}

public final class MediaService extends Service implements MediaPlayer.OnErrorListener,
        MediaPlayer.OnPreparedListener, IMediaService, MediaPlayer.OnCompletionListener {

    @Override
    public void onCompletion(MediaPlayer mp) {
        if (mListener != null) {
            mListener.onCompletion();
        }
    }

    public interface IPlaybackListener {
        void onStartPlayback();
        void onStopPlayback();
        void onCompletion();
    }

    public static final int THIRTY_SECONDS_IN_MILLISECONDS = 30000;
    @NonNull
    public static final String INTENT_EPISODE_URL = "intent_episode_url";
    @NonNull
    public static final String INTENT_EPISODE_TITLE = "intent_episode_title";
    @NonNull
    public static final String INTENT_EPISODE_DESCRIPTION = "intent_episode_description";
    @NonNull
    public static final String ACTION_PLAY = "com.kevintcoughlin.smodr.app.PLAY";
    @NonNull
    public static final String ACTION_PAUSE = "com.kevintcoughlin.smodr.app.PAUSE";
    @NonNull
    public static final String ACTION_RESUME = "com.kevintcoughlin.smodr.app.RESUME";
    @NonNull
    public static final String ACTION_STOP = "com.kevintcoughlin.smodr.app.STOP";
    @NonNull
    public static final String ACTION_FORWARD = "com.kevintcoughlin.smodr.app.FORWARD";
    @NonNull
    public static final String ACTION_REWIND = "com.kevintcoughlin.smodr.app.REWIND";
    @Nullable
    private MediaPlayer mMediaPlayer;
    private final IBinder mBinder = new MediaServiceBinder();
    @Nullable
    private IPlaybackListener mListener;

    public class MediaServiceBinder extends Binder {
        public MediaService getService() {
            return MediaService.this;
        }
    }

    @Contract(pure = true)
    @Override
    public IBinder onBind(Intent intent) {
        return mBinder;
    }

    @NonNull
    public static Intent createIntent(@NonNull final Context context, @NonNull final Item item) {
        final Intent intent = new Intent(context, MediaService.class);
        final String mediaUrlString = item.getUri().toString();

        intent.setAction(MediaService.ACTION_PLAY);
        intent.putExtra(MediaService.INTENT_EPISODE_URL, mediaUrlString);
        intent.putExtra(MediaService.INTENT_EPISODE_TITLE, item.getTitle());
        intent.putExtra(MediaService.INTENT_EPISODE_DESCRIPTION, item.getDescription());

        return intent;
    }

    public int onStartCommand(final Intent intent, final int flags, final int startId) {
        super.onStartCommand(intent, flags, startId);

        final @Nullable String url = intent.getStringExtra(INTENT_EPISODE_URL);
        final String action = intent.getAction();

        if (action != null) {
            switch (action) {
                case ACTION_PAUSE:
                    pausePlayback();
                    break;
                case ACTION_PLAY:
                    startPlayback(url);
                    break;
                case ACTION_RESUME:
                    resumePlayback();
                    break;
                case ACTION_STOP:
                    stopPlayback();
                    break;
                case ACTION_FORWARD:
                    rewind();
                    break;
                case ACTION_REWIND:
                    forward();
                    break;
            }
        }

        return Service.START_REDELIVER_INTENT;
    }

    @Override
    public boolean onError(MediaPlayer mp, int what, int extra) {
        stopPlayback();
        return true;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        stopPlayback();
        if (mMediaPlayer != null) {
            mMediaPlayer.release();
        }
        mListener = null;
    }

    public void seekTo(final int milliseconds) {
        if (mMediaPlayer != null) {
            mMediaPlayer.seekTo(milliseconds);
        }
    }

    public boolean isPlaying() {
        return mMediaPlayer != null && mMediaPlayer.isPlaying();
    }

    @Override
    public int getDurationInMilliseconds() {
        return mMediaPlayer != null ? mMediaPlayer.getDuration() : -1;
    }

    @Override
    public int currentPositionInMilliseconds() {
        return mMediaPlayer != null ? mMediaPlayer.getCurrentPosition() : -1;
    }

    public void resumePlayback() {
        if (mMediaPlayer != null) {
            mMediaPlayer.start();
            if (mListener != null) {
                mListener.onStartPlayback();
            }
        }
    }

    public void pausePlayback() {
        if (mMediaPlayer != null) {
            mMediaPlayer.pause();
            if (mListener != null) {
                mListener.onStopPlayback();
            }
        }
    }

    public void stopPlayback() {
        if (mMediaPlayer != null) {
            mMediaPlayer.stop();
            if (mListener != null) {
                mListener.onStopPlayback();
            }
        }
    }

    @Override
    public void forward() {
        if (mMediaPlayer != null) {
            seekTo(mMediaPlayer.getCurrentPosition() + THIRTY_SECONDS_IN_MILLISECONDS);
        }
    }

    @Override
    public void rewind() {
        if (mMediaPlayer != null) {
            seekTo(mMediaPlayer.getCurrentPosition() - THIRTY_SECONDS_IN_MILLISECONDS);
        }
    }

    public void setPlaybackListener(final IPlaybackListener listener) {
        this.mListener = listener;
    }

    @Override
    public void onPrepared(final MediaPlayer mediaPlayer) {
        resumePlayback();
    }

    private void startPlayback(@Nullable final String url) {
        if (mMediaPlayer != null && mMediaPlayer.isPlaying()) {
            stopPlayback();
        }

        try {
            final Uri uri = Uri.parse(url);

            mMediaPlayer = MediaPlayer.create(this, uri);
            mMediaPlayer.setOnCompletionListener(this);
            mMediaPlayer.start();
            if (mListener != null) {
                mListener.onStartPlayback();
            }
        } catch (NullPointerException exception) {
            Toast.makeText(this, exception.getMessage(), Toast.LENGTH_SHORT).show();
        }
    }
}

File: Smodr/src/main/java/com/kevintcoughlin/smodr/services/MediaService.kt
package com.kevintcoughlin.smodr.services

import android.app.Service
import android.content.Intent
import android.media.MediaPlayer
import android.net.Uri
import android.os.Binder
import android.os.IBinder
import android.util.Log
import androidx.core.net.toUri
import kotlin.math.min

/**
 * Interface defining media service operations.
 */
interface IMediaService {
    fun seekTo(milliseconds: Int)
    val isPlaying: Boolean
    val duration: Int
    val currentTime: Int
    val remainingTime: Int
    fun resumePlayback()
    fun pausePlayback()
    fun stopPlayback()
    fun forward()
    fun rewind()
    fun setPlaybackListener(listener: MediaService.IPlaybackListener?)
}

/**
 * Service for managing media playback with MediaPlayer.
 * Implements playback controls, seeking, and listener notifications.
 */
class MediaService : Service(), MediaPlayer.OnErrorListener, MediaPlayer.OnPreparedListener,
    MediaPlayer.OnCompletionListener, IMediaService {

    /**
     * Interface for playback lifecycle callbacks.
     */
    interface IPlaybackListener {
        fun onStartPlayback()
        fun onStopPlayback()
        fun onCompletion()
    }

    private val mediaPlayer: MediaPlayer by lazy {
        MediaPlayer().apply {
            setOnCompletionListener(this@MediaService)
            setOnErrorListener(this@MediaService)
            setOnPreparedListener(this@MediaService)
        }
    }
    private val binder = MediaServiceBinder()
    private var playbackListener: IPlaybackListener? = null

    inner class MediaServiceBinder : Binder() {
        val service: MediaService
            get() = this@MediaService
    }

    override val currentTime: Int
        get() = mediaPlayer.currentPosition

    override val remainingTime: Int
        get() = (duration - currentTime).coerceAtLeast(0)

    override fun onBind(intent: Intent): IBinder = binder

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        val url = intent.getStringExtra(INTENT_EPISODE_URL)?.toUri()
        handleAction(intent.action, url)
        return START_REDELIVER_INTENT
    }

    private fun handleAction(action: String?, url: Uri?) {
        when (action) {
            ACTION_PAUSE -> pausePlayback()
            ACTION_PLAY -> startPlayback(url)
            ACTION_RESUME -> resumePlayback()
            ACTION_STOP -> stopPlayback()
            ACTION_FORWARD -> forward()
            ACTION_REWIND -> rewind()
        }
    }

    override fun onDestroy() {
        stopPlayback()
        mediaPlayer.release()
        playbackListener = null
        super.onDestroy()
    }

    // IMediaService Implementation

    override val isPlaying: Boolean
        get() = mediaPlayer.isPlaying

    override val duration: Int
        get() = mediaPlayer.duration

    override fun seekTo(milliseconds: Int) {
        mediaPlayer.seekTo(milliseconds)
    }

    override fun resumePlayback() {
        mediaPlayer.start()
        notifyListener { onStartPlayback() }
    }

    override fun pausePlayback() {
        mediaPlayer.pause()
        notifyListener { onStopPlayback() }
    }

    override fun stopPlayback() {
        mediaPlayer.stop()
        notifyListener { onStopPlayback() }
    }

    override fun forward() {
        mediaPlayer.run {
            val newPosition = min(currentPosition + SEEK_INTERVAL, duration)
            seekTo(newPosition)
        }
    }

    override fun rewind() {
        mediaPlayer.run {
            val newPosition = (currentPosition - SEEK_INTERVAL).coerceAtLeast(0)
            seekTo(newPosition)
        }
    }

    override fun setPlaybackListener(listener: IPlaybackListener?) {
        playbackListener = listener
    }

    override fun onCompletion(mp: MediaPlayer) {
        notifyListener { onCompletion() }
    }

    override fun onError(mp: MediaPlayer, what: Int, extra: Int): Boolean {
        stopPlayback()
        return true
    }

    override fun onPrepared(mp: MediaPlayer) {
        resumePlayback()
    }

    /**
     * Starts playback of media from the given URL.
     *
     * @param url The URI of the media to play
     */
    fun startPlayback(url: Uri?) {
        url ?: return
        try {
            if (mediaPlayer.isPlaying) stopPlayback()
            mediaPlayer.reset()
            mediaPlayer.setDataSource(this, url)
            mediaPlayer.prepareAsync()
        } catch (e: Exception) {
            Log.e(TAG, "Error starting playback: ${e.message}", e)
        }
    }

    private inline fun notifyListener(action: IPlaybackListener.() -> Unit) {
        playbackListener?.action()
    }

    companion object {
        const val SEEK_INTERVAL = 30_000 // 30 seconds in milliseconds
        const val INTENT_EPISODE_URL = "intent_episode_url"
        const val ACTION_PLAY = "com.kevintcoughlin.smodr.app.PLAY"
        const val ACTION_PAUSE = "com.kevintcoughlin.smodr.app.PAUSE"
        const val ACTION_RESUME = "com.kevintcoughlin.smodr.app.RESUME"
        const val ACTION_STOP = "com.kevintcoughlin.smodr.app.STOP"
        const val ACTION_FORWARD = "com.kevintcoughlin.smodr.app.FORWARD"
        const val ACTION_REWIND = "com.kevintcoughlin.smodr.app.REWIND"
        private const val TAG = "MediaService"
    }
}
